command/e2etest/init_test.go

```diff
@@ -375,6 +379,57 @@ func TestInitProviderNotFound(t *testing.T) {
 │ Could not retrieve the list of available versions for provider
 │ hashicorp/nonexist: provider registry registry.terraform.io does not have a
 │ provider named registry.terraform.io/hashicorp/nonexist
+│ 
+│ If you meant to install a provider from the registry that is not in the
+│ hashicorp namespace, the required_providers block for your module is not
+│ optional. All modules should specify their required_providers so that
+│ external consumers will get the correct providers when using a module.
```

Comment:
Self comment: this should mention `terraform providers` somewhere so people can use it to see where their provider needs are coming from.